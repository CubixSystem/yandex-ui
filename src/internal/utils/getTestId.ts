/**
 * Возвращает ID для использования в тестах.
 *
 * @example
 *
 * getTestId('host', testId)
 */
export function getTestId(prefix:string, id?: string): string | undefined {
    if (id === undefined) {
        return undefined;
    }
    return `${id}-${prefix}`;
}
