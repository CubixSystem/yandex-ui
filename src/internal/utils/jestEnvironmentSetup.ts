/**
 * Возвращает DOM-узел для монтирования компонента.
 */
export const getRoot = (): HTMLElement | null => {
    return document.getElementById('root');
};

/**
 * Хелпер, создающий корневой контейнер для монтирования,
 * необходим для того, чтобы избежать ошибок при монтировании в enzyme.
 */
export const jestEnvironmentSetup = (): void => {
    // eslint-disable-next-line no-undef
    beforeEach(() => {
        const node = document.createElement('div');
        node.setAttribute('id', 'root');
        document.body.appendChild(node);
    });

    // eslint-disable-next-line no-undef
    afterEach(() => {
        const node = getRoot();
        if (node !== null) {
            document.body.removeChild(node);
        }
    });
};
