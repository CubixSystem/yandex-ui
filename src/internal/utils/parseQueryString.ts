function parseValue(value: any) {
    if (!isNaN(Number(value)) && (typeof value === 'string' && value.trim() !== '')) {
        value = Number(value);
    } else if (value !== null && (value.toLowerCase() === 'true' || value.toLowerCase() === 'false')) {
        value = value.toLowerCase() === 'true';
    }
    return value;
}

export function parseQueryString<T>(search: string): T {
    const hashes = search.slice(search.indexOf('?') + 1).split('&');
    return hashes.reduce(
        (params, hash) => {
            const split = hash.indexOf('=');
            const key = hash.slice(0, split);
            const value = parseValue(hash.slice(split + 1));
            return Object.assign(params, { [key]: value });
        },
        {} as any,
    );
}
