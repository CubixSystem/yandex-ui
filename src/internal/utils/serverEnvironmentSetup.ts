/**
 * Хелпер, который симулирует серверное окружение во время запуска тестов.
 */
export const serverEnvironmentSetup = () => {
    const prevWindow = (global as any).window;
    const prevDocument = (global as any).document;
    const prevRAF = (global as any).requestAnimationFrame;

    // eslint-disable-next-line no-undef
    beforeEach(() => {
        delete (global as any).window;
        delete (global as any).document;
        delete (global as any).requestAnimationFrame;
    });

    // eslint-disable-next-line no-undef
    afterEach(() => {
        (global as any).window = prevWindow;
        (global as any).document = prevDocument;
        (global as any).requestAnimationFrame = prevRAF;
    });
};
