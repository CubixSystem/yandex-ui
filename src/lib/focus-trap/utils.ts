import { getTabbables } from './tabbable';

const FOCUS_TRAP_GUARD_ATTR = 'focusTrapGuard';

interface ScrollableElement {
    element: HTMLElement;
    scrollTop: number;
    scrollLeft: number;
}

let supportsPreventScrollCached: boolean | null = null;
function supportsPreventScroll() {
    if (supportsPreventScrollCached === null) {
        supportsPreventScrollCached = false;

        try {
            const element = document.createElement('div');

            element.focus({
                get preventScroll() {
                    supportsPreventScrollCached = true;
                    return true;
                },
            });
        } catch (e) {
            // ignore
        }
    }

    return supportsPreventScrollCached;
}

function getScrollableElements(element: HTMLElement): ScrollableElement[] {
    const scrollableElements: ScrollableElement[] = [];
    const rootScrollingElement = document.scrollingElement || document.documentElement;

    let parent = element.parentNode;
    while (parent instanceof HTMLElement && parent !== rootScrollingElement) {
        if (parent.offsetHeight < parent.scrollHeight || parent.offsetWidth < parent.scrollWidth) {
            scrollableElements.push({
                element: parent,
                scrollTop: parent.scrollTop,
                scrollLeft: parent.scrollLeft,
            });
        }
        parent = parent.parentNode;
    }

    if (rootScrollingElement instanceof HTMLElement) {
        scrollableElements.push({
            element: rootScrollingElement,
            scrollTop: rootScrollingElement.scrollTop,
            scrollLeft: rootScrollingElement.scrollLeft,
        });
    }

    return scrollableElements;
}

function restoreScrollPosition(scrollableElements: ScrollableElement[]) {
    for (const { element, scrollTop, scrollLeft } of scrollableElements) {
        element.scrollTop = scrollTop;
        element.scrollLeft = scrollLeft;
    }
}

export function getActiveElement() {
    return document.activeElement as HTMLElement | null;
}

export function isElementInScope(scope: Element, element: Element) {
    return scope.contains(element);
}

export function isFocusFree() {
    return getActiveElement() === document.body;
}

export function createGuard(tabIndex: number) {
    const element = document.createElement('span');
    element.tabIndex = tabIndex;
    Object.assign(element.style, {
        width: '1px',
        height: 0,
        padding: 0,
        overflow: 'hidden',
        position: 'fixed',
        top: '1px',
        left: '1px',
    });
    element.setAttribute('aria-hidden', 'true');
    element.dataset[FOCUS_TRAP_GUARD_ATTR] = 'true';

    return element;
}

export function isGuard(node: HTMLElement) {
    return Boolean(node.dataset[FOCUS_TRAP_GUARD_ATTR]);
}

export function removeElementFromDocument(element: Element) {
    if (!element.parentNode) {
        return;
    }

    element.parentNode.removeChild(element);
}

function focusWithoutScrolling(element: HTMLElement) {
    if (supportsPreventScroll()) {
        element.focus({ preventScroll: true });
    } else {
        const scrollableElements = getScrollableElements(element);
        element.focus();
        restoreScrollPosition(scrollableElements);
    }
}

export function focusElement(element?: HTMLElement | null, preventScroll?: boolean) {
    if (!element) {
        return;
    }

    try {
        if (preventScroll) {
            focusWithoutScrolling(element);
        } else {
            element.focus();
        }
    } catch (e) {
        // ignore
    }
}

export function focusFirstIn(scope: HTMLElement) {
    const [first] = getTabbables(scope, (node) => !isGuard(node));

    focusElement(first);
}

export function getNextTabbableIndex(current: number, count: number, step: number) {
    if (current === -1) {
        return 0;
    }

    return count > 0 ? (count + current + step) % count : -1;
}
