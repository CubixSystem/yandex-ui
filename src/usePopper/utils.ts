import { Placement, BasePlacement, Offsets, top, left, right } from '@popperjs/core';

export function getLayoutRect(element: HTMLElement) {
    return {
        x: element.offsetLeft,
        y: element.offsetTop,
        width: element.offsetWidth,
        height: element.offsetHeight,
    };
}

export default function getBasePlacement(placement: Placement): BasePlacement {
    return placement.split('-')[0] as any;
}

export function distanceAndSkiddingToXY(placement: Placement, offset: [number, number]): Offsets {
    const basePlacement = getBasePlacement(placement);
    const invertDistance = [left, top].indexOf(basePlacement as any) >= 0 ? -1 : 1;

    const skidding = offset[0] || 0;
    const distance = (offset[1] || 0) * invertDistance;

    return [left, right].indexOf(basePlacement as any) >= 0
        ? { x: distance, y: skidding }
        : { x: skidding, y: distance };
}
