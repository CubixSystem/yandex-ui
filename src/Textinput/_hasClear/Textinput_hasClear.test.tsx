import React, { createRef, ComponentType } from 'react';
import { mount, render } from 'enzyme';
import { ExtractProps } from '@bem-react/core';

import { serverEnvironmentSetup, delay } from '../../internal/utils';
import { jestEnvironmentSetup, getRoot } from '../../internal/utils/jestEnvironmentSetup';

import { Textinput as TextinputDesktop } from '../Textinput@desktop';
import { Textinput as TextinputTouchPad } from '../Textinput@touch-pad';
import { Textinput as TextinputTouchPhone } from '../Textinput@touch-phone';
import { withHasClear as withHasClearDesktop } from './Textinput_hasClear@desktop';
import { withHasClear as withHasClearTouchPad } from './Textinput_hasClear@touch-pad';
import { withHasClear as withHasClearTouchPhone } from './Textinput_hasClear@touch-phone';

const platforms = [
    ['desktop', withHasClearDesktop(TextinputDesktop)],
    ['touch-pad', withHasClearTouchPad(TextinputTouchPad)],
    ['touch-phone', withHasClearTouchPhone(TextinputTouchPhone)],
];

const TextinputForType = withHasClearDesktop(TextinputDesktop);
type TextinputProps = ExtractProps<typeof TextinputForType>;

describe.each<any>(platforms)('Textinput_hasClear@%s', (_platform, Textinput: ComponentType<TextinputProps>) => {
    jestEnvironmentSetup();

    describe('server environment', () => {
        serverEnvironmentSetup();

        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(render(<Textinput hasClear />)).toMatchSnapshot();
        });
    });

    describe('client environment', () => {
        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(render(<Textinput hasClear />)).toMatchSnapshot();
        });

        test('должен вызывать событие onChange с пустым значением в event при клике в clear', () => {
            const onChange = jest.fn();
            const wrapper = mount(<Textinput hasClear onChange={onChange} />);
            wrapper
                .find('.Textinput-Clear')
                .first()
                .simulate('click');
            expect(onChange).toHaveBeenCalledTimes(1);
            expect(onChange.mock.calls[0][0].target.value).toBe('');
        });

        test('должен вызывать событие onClearClick при клике в clear', () => {
            const onClearClick = jest.fn();
            const wrapper = mount(<Textinput hasClear onClearClick={onClearClick} />);
            wrapper
                .find('.Textinput-Clear')
                .first()
                .simulate('click');
            expect(onClearClick).toHaveBeenCalledTimes(1);
        });

        test('должен проставлять фокус в input после клика в clear', async () => {
            const controlRef = createRef<HTMLInputElement>();
            const wrapper = mount(<Textinput hasClear controlRef={controlRef} />, { attachTo: getRoot() });
            await delay(100);
            wrapper
                .find('.Textinput-Clear')
                .first()
                .simulate('click');
            expect(controlRef.current).toEqual(document.activeElement);
        });
    });
});
