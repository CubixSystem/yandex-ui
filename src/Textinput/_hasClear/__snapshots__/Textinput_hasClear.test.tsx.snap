// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textinput_hasClear@desktop client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textinput Textinput_hasClear"
>
  <span
    aria-hidden="true"
    class="Icon Icon_type_cross Textinput-Icon Textinput-Clear"
  />
  <input
    aria-invalid="false"
    autocomplete="off"
    class="Textinput-Control"
  />
  <span
    class="Textinput-Box"
  />
</span>
`;

exports[`Textinput_hasClear@desktop server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textinput Textinput_hasClear"
>
  <span
    aria-hidden="true"
    class="Icon Icon_type_cross Textinput-Icon Textinput-Clear"
  />
  <input
    aria-invalid="false"
    autocomplete="off"
    class="Textinput-Control"
  />
  <span
    class="Textinput-Box"
  />
</span>
`;

exports[`Textinput_hasClear@touch-pad client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textinput Textinput_hasClear"
>
  <span
    aria-hidden="true"
    class="Icon Icon_type_cross Textinput-Icon Textinput-Clear"
  />
  <input
    aria-invalid="false"
    autocomplete="off"
    class="Textinput-Control"
  />
  <span
    class="Textinput-Box"
  />
</span>
`;

exports[`Textinput_hasClear@touch-pad server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textinput Textinput_hasClear"
>
  <span
    aria-hidden="true"
    class="Icon Icon_type_cross Textinput-Icon Textinput-Clear"
  />
  <input
    aria-invalid="false"
    autocomplete="off"
    class="Textinput-Control"
  />
  <span
    class="Textinput-Box"
  />
</span>
`;

exports[`Textinput_hasClear@touch-phone client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textinput Textinput_hasClear"
>
  <span
    aria-hidden="true"
    class="Icon Icon_type_cross Textinput-Icon Textinput-Clear"
  />
  <input
    aria-invalid="false"
    autocomplete="off"
    class="Textinput-Control"
  />
  <span
    class="Textinput-Box"
  />
</span>
`;

exports[`Textinput_hasClear@touch-phone server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textinput Textinput_hasClear"
>
  <span
    aria-hidden="true"
    class="Icon Icon_type_cross Textinput-Icon Textinput-Clear"
  />
  <input
    aria-invalid="false"
    autocomplete="off"
    class="Textinput-Control"
  />
  <span
    class="Textinput-Box"
  />
</span>
`;
